### PR TITLE
stabilized compiler_fences (fixes #41091)

### LINKED ISSUE #41091: Provide safe wrappers for single-threaded memory fence intrinsics
[RFC #888](https://github.com/rust-lang/rfcs/pull/888) introduced single-threaded (i.e., compiler-only) memory fence intrinsics, and was implemented in https://github.com/rust-lang/rust/pull/24833.

The RFC explicitly does not add safe wrappers for these new barriers, and states
> The existing fence intrinsics are exported in libstd with safe wrappers, but this design does not export safe wrappers for the new intrinsics. The existing fence functions will still perform correctly if used where a single-threaded fence is called for, but with a slight reduction in efficiency. Not exposing these new intrinsics through a safe wrapper reduces the possibility for confusion on which fences are appropriate in a given situation, while still providing the capability for users to opt in to a single-threaded fence when appropriate.

While the argument is sound, authors of low-level concurrency libraries are often very concerned about the performance overhead of full-blown memory fences when a simple compiler barrier will do. The current design requires authors of such libraries to either require a nightly compiler (to use intrinsics), or to fall back to the old
```
asm!("" ::: "memory" : "volatile")
```
workaround (which for the time being also requires nightly due to `feature(asm)`).

I propose adding in a safe wrapper around these compiler barrier intrinsics alongside `fence` after all, with name and documentation that clearly indicates how that function differs from what `fence` provides. PR incoming shortly.

### PR DESCRIPTION
**Pull request by budziq**
I did not know what to proceed with "unstable-book" entry. The feature would no longer be unstable so I have deleted it. If it was the wrong call I'll revert it (unfortunately his case is not described in the CONTRIBUTING.md).

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! Maybe the docs could move to the function?

**Comment by budziq** [comment]
> Maybe the docs could move to the function?

I've translated part of the unstable-book article into an example. I'm not sure if it didn't came out forced.

**Comment by alexcrichton** [comment]
@bors: r+

Sure!

cc @jonhoo

**Comment by jonhoo** [comment]
Hmm, I think I would argue for more of the text from the unstable book to be incorporated into the documentation for `compiler_fence`. This is a very specialized function, and having the docs to explain that is probably a good idea. In particular, I think the last paragraph before the Examples header, and the first paragraph after, could be kept verbatim in the docs for `compiler_fence`, perhaps under a heading of "Usage"? There's also the question about whether it is appropriate to refer readers to the equivalent function in C++ like the unstable book text does, though I feel less strongly about that.

**Comment by alexcrichton** [comment]
@bors: r-

@budziq want to update with @jonhoo's thoughts?

**Comment by budziq** [comment]
> want to update with @jonhoo's thoughts?

@alexcrichton  Gladly! Although I might not have time to approach it this weekend.

> In particular, I think the last paragraph before the Examples header, and the first paragraph after, could be kept verbatim

@jonhoo I've originally omitted these paragraphs as these convey the same message as current docs present above `compiler_fence` (although in much terser form). I can replace one with the other.

 > There's also the question about whether it is appropriate to refer readers to the equivalent function in C++

I'm not sure about linking to C++ reference but I would prefer not to point to StackOverflow, leaving only the linux kernel link.

**Comment by budziq** [comment]
I've updated the docs with the best I could come up with :)

**Comment by jonhoo** [comment]
LGTM :+1:

File: src/doc/unstable-book/src/library-features/compiler-fences.md [hunk 1]
```diff
@@ -1,106 +0,0 @@
...
-    if IS_READY.load(Ordering::Relaxed) {
-        assert_eq!(IMPORTANT_VARIABLE.load(Ordering::Relaxed), 42);
-    }
-}
-```
-
-The way it is currently written, the `assert_eq!` is *not* guaranteed to
-succeed, despite everything happening in a single thread. To see why,
-remember that the compiler is free to swap the stores to
-`IMPORTANT_VARIABLE` and `IS_READ` since they are both
```
Comment (nalply) [inline src/doc/unstable-book/src/library-features/compiler-fences.md]: Perhaps a typo? IS_READY instead of IS_READ

Comment (jonhoo) [reply src/doc/unstable-book/src/library-features/compiler-fences.md]: Yes, this should be `IS_READY`

**Pull request merged**